File: apps/reading/src/api/collectiveDeepResearchMerge.ts
/**
 * Collective deep research → written analysis draft intent (pure).
 *
 * Operator vision: after multiple floating/sub-agent deep researches complete,
 * merge them into a written analysis draft (or full authorize with ack).
 *
 * analysis_written always false in pure layer — intent only.
 */

export type AnalysisMergeKind = "draft_analysis" | "full_analysis";

export interface CompletedResearchInstance {
  instance_id: string;
  parent_asset_id: string;
  status: "completed" | "open" | "proposed" | "closed";
  highlight?: string;
  prompt?: string;
  /** Optional operator-supplied findings — never invented. */
  findings?: string[];
}

export interface CollectiveAnalysisIntent {
  kind: AnalysisMergeKind;
  parent_asset_id: string;
  instance_ids: string[];
  findings: string[];
  operator_ack: boolean;
  /** Always false — pure intent does not write analysis asset. */
  analysis_written: false;
  notes: string[];
  authority: "collective_analysis_intent_only";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Propose merging ≥2 completed (or open) floating research instances into a
 * written analysis draft/full intent. Never invents findings; never writes.
 */
export function proposeCollectiveAnalysisMerge(
  instances: CompletedResearchInstance[],
  input: {
    kind: AnalysisMergeKind;
    operator_ack: boolean;
    /** Optional additional operator findings. */
    extra_findings?: string[] | null;
  },
): CollectiveAnalysisIntent {
  if (!Array.isArray(instances) || instances.length < 2) {
    throw new Error("collective analysis requires at least 2 instances");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (input.kind !== "draft_analysis" && input.kind !== "full_analysis") {
    throw new Error("kind must be draft_analysis or full_analysis");
  }
  if (input.kind === "full_analysis" && input.operator_ack !== true) {
    throw new Error("full_analysis requires operator_ack=true (fail closed)");
  }

  const parent = requireNonEmpty(
    instances[0]?.parent_asset_id,
    "instances[0].parent_asset_id",
  );
  const ids: string[] = [];
  const findings: string[] = [];
  const notes: string[] = [];

  for (let i = 0; i < instances.length; i++) {
    const inst = instances[i];
    if (!inst || typeof inst !== "object") {
      throw new Error(`instances[${i}] must be an object`);
    }
    const id = requireNonEmpty(inst.instance_id, `instances[${i}].instance_id`);
    const p = requireNonEmpty(
      inst.parent_asset_id,
      `instances[${i}].parent_asset_id`,
    );
    if (p !== parent) {
      throw new Error("collective analysis requires same parent_asset_id");
    }
    if (
      inst.status !== "completed" &&
      inst.status !== "open" &&
      inst.status !== "proposed"
    ) {
      throw new Error(
        `instances[${i}] status must be proposed|open|completed (not closed)`,
      );
    }
    if (input.kind === "full_analysis" && inst.status !== "completed") {
      throw new Error("full_analysis requires all instances completed");
    }
    ids.push(id);
    if (inst.findings != null) {
      if (!Array.isArray(inst.findings)) {
        throw new Error(`instances[${i}].findings must be string[] when set`);
      }
      for (let j = 0; j < inst.findings.length; j++) {
        const f = inst.findings[j];
        if (typeof f !== "string" || !f.trim()) {
          throw new Error(`instances[${i}].findings[${j}] must be non-empty string`);
        }
        findings.push(f.trim());
      }
    }
  }

  // de-dupe instance ids
  const seen = new Set<string>();
  const unique = ids.filter((id) => {
    if (seen.has(id)) return false;
    seen.add(id);
    return true;
  });
  if (unique.length < 2) {
    throw new Error("collective analysis requires at least 2 distinct instance_ids");
  }

  if (input.extra_findings != null) {
    if (!Array.isArray(input.extra_findings)) {
      throw new Error("extra_findings must be string[] or null");
    }
    for (let j = 0; j < input.extra_findings.length; j++) {
      const f = input.extra_findings[j];
      if (typeof f !== "string" || !f.trim()) {
        throw new Error(`extra_findings[${j}] must be non-empty string`);
      }
      findings.push(f.trim());
    }
  }

  if (findings.length === 0) {
    notes.push(
      "no findings supplied — analysis scaffold intent only (no invent content)",
    );
  } else {
    notes.push(`findings=${findings.length} caller-supplied only`);
  }

  notes.push("analysis_written=false");
  notes.push(
    input.kind === "draft_analysis"
      ? "draft analysis intent — provisional combined document not written"
      : "full analysis intent — parent/analysis asset not mutated in pure layer",
  );

  return {
    kind: input.kind,
    parent_asset_id: parent,
    instance_ids: unique,
    findings,
    operator_ack: input.operator_ack,
    analysis_written: false,
    notes,
    authority: "collective_analysis_intent_only",
  };
}

export function formatCollectiveAnalysisSummary(
  intent: CollectiveAnalysisIntent,
): string {
  return (
    `${intent.kind} · instances=${intent.instance_ids.length} · ` +
    `findings=${intent.findings.length} · analysis_written=false`
  );
}

File: substrate/collective_deep_research_merge.py
"""Collective deep research → written analysis intent (pure).

Merges ≥2 floating/sub-agent research instances into draft/full analysis
intent. analysis_written is always False. Findings are caller-supplied only.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

AnalysisMergeKind = Literal["draft_analysis", "full_analysis"]
InstanceStatus = Literal["completed", "open", "proposed", "closed"]


class CollectiveAnalysisMergeError(ValueError):
    """Fail-closed validation for collective analysis merge."""


@dataclass(frozen=True)
class CollectiveAnalysisIntent:
    kind: AnalysisMergeKind
    parent_asset_id: str
    instance_ids: tuple[str, ...]
    findings: tuple[str, ...]
    operator_ack: bool
    analysis_written: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "kind": self.kind,
            "parent_asset_id": self.parent_asset_id,
            "instance_ids": list(self.instance_ids),
            "findings": list(self.findings),
            "operator_ack": self.operator_ack,
            "analysis_written": False,
            "notes": list(self.notes),
            "authority": "collective_analysis_intent_only",
        }


def _require_bool(value: object, *, field: str) -> bool:
    if not isinstance(value, bool):
        raise CollectiveAnalysisMergeError(f"{field} must be an explicit boolean")
    return value


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise CollectiveAnalysisMergeError(f"{field} must be a non-empty string")
    return value.strip()


def propose_collective_analysis_merge(
    instances: object,
    *,
    kind: object,
    operator_ack: object,
    extra_findings: object | None = None,
) -> CollectiveAnalysisIntent:
    """Propose draft/full analysis intent. Never writes analysis assets."""
    ack = _require_bool(operator_ack, field="operator_ack")
    if kind not in ("draft_analysis", "full_analysis"):
        raise CollectiveAnalysisMergeError(
            "kind must be draft_analysis or full_analysis"
        )
    kind_s: AnalysisMergeKind = kind
    if kind_s == "full_analysis" and ack is not True:
        raise CollectiveAnalysisMergeError(
            "full_analysis requires operator_ack=true (fail closed)"
        )
    if not isinstance(instances, list) or len(instances) < 2:
        raise CollectiveAnalysisMergeError(
            "collective analysis requires at least 2 instances"
        )

    parent = _require_nonempty(
        instances[0].get("parent_asset_id") if isinstance(instances[0], dict) else None,
        field="instances[0].parent_asset_id",
    )
    ids: list[str] = []
    findings: list[str] = []
    notes: list[str] = []

    for i, inst in enumerate(instances):
        if not isinstance(inst, dict):
            raise CollectiveAnalysisMergeError(f"instances[{i}] must be an object")
        iid = _require_nonempty(inst.get("instance_id"), field=f"instances[{i}].instance_id")
        p = _require_nonempty(
            inst.get("parent_asset_id"), field=f"instances[{i}].parent_asset_id"
        )
        if p != parent:
            raise CollectiveAnalysisMergeError(
                "collective analysis requires same parent_asset_id"
            )
        status = inst.get("status")
        if status not in ("completed", "open", "proposed", "closed"):
            raise CollectiveAnalysisMergeError(
                f"instances[{i}].status must be proposed|open|completed|closed"
            )
        if status == "closed":
            raise CollectiveAnalysisMergeError(
                f"instances[{i}] status must be proposed|open|completed (not closed)"
            )
        if kind_s == "full_analysis" and status != "completed":
            raise CollectiveAnalysisMergeError(
                "full_analysis requires all instances completed"
            )
        ids.append(iid)
        raw_findings = inst.get("findings")
        if raw_findings is not None:
            if not isinstance(raw_findings, list):
                raise CollectiveAnalysisMergeError(
                    f"instances[{i}].findings must be string[] when set"
                )
            for j, f in enumerate(raw_findings):
                if not isinstance(f, str) or not f.strip():
                    raise CollectiveAnalysisMergeError(
                        f"instances[{i}].findings[{j}] must be non-empty string"
                    )
                findings.append(f.strip())

    seen: set[str] = set()
    unique: list[str] = []
    for iid in ids:
        if iid in seen:
            continue
        seen.add(iid)
        unique.append(iid)
    if len(unique) < 2:
        raise CollectiveAnalysisMergeError(
            "collective analysis requires at least 2 distinct instance_ids"
        )

    if extra_findings is not None:
        if not isinstance(extra_findings, list):
            raise CollectiveAnalysisMergeError("extra_findings must be string[] or null")
        for j, f in enumerate(extra_findings):
            if not isinstance(f, str) or not f.strip():
                raise CollectiveAnalysisMergeError(
                    f"extra_findings[{j}] must be non-empty string"
                )
            findings.append(f.strip())

    if not findings:
        notes.append(
            "no findings supplied — analysis scaffold intent only (no invent content)"
        )
    else:
        notes.append(f"findings={len(findings)} caller-supplied only")

    notes.append("analysis_written=false")
    notes.append(
        "draft analysis intent — provisional combined document not written"
        if kind_s == "draft_analysis"
        else "full analysis intent — parent/analysis asset not mutated in pure layer"
    )

    return CollectiveAnalysisIntent(
        kind=kind_s,
        parent_asset_id=parent,
        instance_ids=tuple(unique),
        findings=tuple(findings),
        operator_ack=ack,
        analysis_written=False,
        notes=tuple(notes),
        authority="collective_analysis_intent_only",
    )


__all__ = [
    "CollectiveAnalysisIntent",
    "CollectiveAnalysisMergeError",
    "propose_collective_analysis_merge",
]
